core/embed/rust/src/zcash_primitives/pallas/scalar.rs

```diff
@@ -0,0 +1,108 @@
+use core::ops::Deref;
+
+use pasta_curves::{
+    arithmetic::FieldExt,
+    group::ff::PrimeField,
+    pallas::{Point, Scalar},
+};
+
+use crate::micropython::{
+    ffi,
+    gc::Gc,
+    map::Map,
+    obj::Obj,
+    qstr::Qstr,
+    typ::Type,
+    util,
+    wrap::{Wrapable, Wrapped},
+};
+
+pub static SCALAR_TYPE: Type = obj_type! {
+    name: Qstr::MP_QSTR_Scalar,
+    locals: &obj_dict!(obj_map! {
+            Qstr::MP_QSTR_to_bytes => obj_fn_1!(scalar_to_bytes).as_obj(),
+            Qstr::MP_QSTR_is_not_zero => obj_fn_1!(scalar_is_not_zero).as_obj(),
+    }),
+    make_new_fn: super::common::ff_from_bytes::<Scalar>,
+    unary_op_fn: scalar_unary_op,
+    binary_op_fn: scalar_binary_op,
+};
+
+impl Wrapable for Scalar {
+    fn obj_type() -> &'static Type {
+        &SCALAR_TYPE
+    }
+}
+
+unsafe extern "C" fn scalar_binary_op(op: ffi::mp_binary_op_t, this: Obj, other: Obj) -> Obj {
+    let block = || {
+        let this = Gc::<Wrapped<Scalar>>::try_from(this)?;
+        let this = this.deref().inner();
+        match op {
+            ffi::mp_binary_op_t_MP_BINARY_OP_MULTIPLY => {
+                let other = Gc::<Wrapped<Point>>::try_from(other)?;
+                let other = other.deref().inner();
+                (other * this).wrap()
+            }
+            ffi::mp_binary_op_t_MP_BINARY_OP_ADD => {
+                let other = Gc::<Wrapped<Scalar>>::try_from(other)?;
+                let other = other.deref().inner();
+                (other + this).wrap()
+            }
+            _ => Ok(Obj::const_null()),
+        }
+    };
+    unsafe { util::try_or_raise(block) }
+}
+
+unsafe extern "C" fn scalar_unary_op(op: ffi::mp_binary_op_t, this: Obj) -> Obj {
```

Comment:
dtto the arg should be `mp_unary_op_t`

Reply:
55f900a053f44428a389ef75c1f50ca7ea3ec26b